hs_core/models.py

```diff
@@ -1354,6 +1303,77 @@ class AbstractResource(ResourcePermissionsMixin):
         class MyResourceContentType(pages.Page, hs_core.AbstractResource):
             ...
     """
+    # For each request for a resource page view, the Page will run can_view(), can_change(),
+    # can_delete() in succession to check for permissions, so for performance improvement,
+    # there might be merit to make ha_obj a class member variable with a persistent connection
+    # for the request, rather than a separate connection within each method call
+    def can_add(self, request):
+        return self.can_change(request)
+
+    def can_delete(self, request):
+        user = get_user(request)
+        if user.is_authenticated():
+            if user.is_superuser:
+                ret = True
+            elif self.creator and user.pk == self.creator.pk:
+                ret = True
+            else:
+                try:
+                    ha_obj = HSAlib.HSAccess(str(user.username), 'unused', settings.HS_ACCESS_DB, settings.HS_ACCESS_USERNAME,
+                                             settings.HS_ACCESS_PASSWORD, settings.HS_ACCESS_HOST, settings.HS_ACCESS_PORT)
+                    ret = ha_obj.resource_is_owned(self.short_id)
+                    del ha_obj
+                except:
+                    # unable to connect to the database
+                    ret = False
+        else:
+            ret = False
+        return ret
+
+    def can_change(self, request):
+        user = get_user(request)
+        if user.is_authenticated():
+            if user.is_superuser:
+                ret = True
+            elif self.creator and user.pk == self.creator.pk:
```

Comment:
See long response above about creators and owners.  I am not sure in the code above if the term creator is being used in a sense consistent with the intended access control model.